### PR TITLE
GH-41475: [Python] Build with Python 3.13

### LINKED ISSUE #41475: [Python] _Py_IsFinalizing will be removed in Python 3.13
### Describe the bug, including details regarding any error messages, version, and platform.

`_Py_IsFinalizing` was removed in https://github.com/python/cpython/pull/106400 and replaced with a public version in https://github.com/python/cpython/pull/108032 .

This function is used 5 places in `udf.cc` for example: https://github.com/apache/arrow/blob/0ef7351986ee8b967e210d0f9c7a9c8e4d4038fd/python/pyarrow/src/arrow/python/udf.cc#L49-L53 which prevents building on Python 3.13.

Presumably this will need to be version gated on the version of CPython, but I am not sure how you prefer to do that in arrow.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by tacaswell**
### Rationale for this change

The private function `_Py_IsFinalizing` renamed to `Py_IsFinalizing` in Python 3.13.0a1.  Without this change pyarrow will not compile with this version of Python or higher.

### What changes are included in this PR?

add a version-gated `#define` to handle the change.

### Are these changes tested?

Local build succeeds. 

### Are there any user-facing changes?

no

* GitHub Issue: #41475

### CODE REVIEW

File: python/pyarrow/src/arrow/python/udf.cc [hunk 1]
```diff
@@ -28,6 +28,11 @@
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
 
+#if PY_VERSION_HEX >= 0x030D00A4
+#define __Py_IsFinalizing Py_IsFinalizing
+#else
+#define __Py_IsFinalizing _Py_IsFinalizing
+#endif
```
Comment (kou) [inline python/pyarrow/src/arrow/python/udf.cc:35]: How about always using `Py_IsFinalizing` (new API)?

```suggestion
#if PY_VERSION_HEX < 0x030D00A4
#define Py_IsFinalizing() _Py_IsFinalizing()
#endif
```

**Comment by pitrou** [comment]
Instead of doing such adjustements one by one, how about we vendor a version of https://github.com/python/pythoncapi-compat ? cc @jorisvandenbossche

File: python/pyarrow/src/arrow/python/udf.cc (hunk 1 above)
Comment (tacaswell) [reply python/pyarrow/src/arrow/python/udf.cc:35]: Fair, I'll merge this, and then amend in the rest of the changes.

**Comment by jorisvandenbossche** [comment]
> Instead of doing such adjustements one by one, how about we vendor a version of https://github.com/python/pythoncapi-compat ?

I am not very familiar with it to estimate how much we would make use of it (this PR also essentially only adds 3 lines, and the actual changes to change `_Py_IsFinalizing` to `Py_IsFinalizing` is still something that we would had to do). But yes, if that tool is exactly meant for this, maybe why not just use it.

I quickly tested its upgrade script on the pyarrow C++ code, and what it does change is replace a bunch of `.. == Py_None` cases with `Py_IsNone(..)`.

But maybe the more interesting usage of the tool would be that we can use newer C API functions in our code that might make our PyArrow C++ easier to read or maintain? (but again not sure how much of those cases would actually apply here)

File: python/pyarrow/src/arrow/python/udf.cc [hunk 2]
```diff
@@ -28,6 +28,9 @@
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
 
+#if PY_VERSION_HEX < 0x030D00A4
```
Comment (jorisvandenbossche) [inline python/pyarrow/src/arrow/python/udf.cc:31]:
```suggestion
// Py_IsFinalizing added in Python 3.13.0a4
#if PY_VERSION_HEX < 0x030D00A4
```

**Comment by jorisvandenbossche** [comment]
(I am also fine with considering https://github.com/python/pythoncapi-compat for a follow-up issue, and at least ensure now pyarrow builds with Python 3.13)

**Comment by pitrou** [comment]
> I quickly tested its upgrade script on the pyarrow C++ code, and what it does change is replace a bunch of `.. == Py_None` cases with `Py_IsNone(..)`.
> 
> But maybe the more interesting usage of the tool would be that we can use newer C API functions in our code that might make our PyArrow C++ easier to read or maintain? (but again not sure how much of those cases would actually apply here)

Yes, the idea would be to use newer C API functions as in this PR. Vendoring it should be trivial: there is [a single `.h` file](https://github.com/python/pythoncapi-compat/blob/main/pythoncapi_compat.h).

**Comment by jorisvandenbossche** [comment]
I don't have time this or coming week to look into vendoring [pythoncapi-compat](https://github.com/python/pythoncapi-compat), ad given the upcoming feature freeze for 17.0 the week after, I would suggest we merge this for now and track using pythoncapi-compat as a follow-up issue.

**Review by kou — APPROVED** [review]
+1

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python

**Review by raulcd — APPROVED** [review]
I am running extra CI for this. If tests are successful I'll merge as is and we can open a follow up issue to vendor [pythoncapi-compat](https://github.com/python/pythoncapi-compat) as suggested

**Comment by raulcd** [comment]
Failures are unrelated

**Comment by raulcd** [comment]
I've created the follow up issue https://github.com/apache/arrow/issues/43069 to vendor https://github.com/python/pythoncapi-compat

**Pull request merged**